crates/oxc_transformer/src/react/mod.rs

```diff
@@ -78,8 +78,29 @@ impl<'a> React<'a> {
         }
     }
 
-    pub fn transform_jsx_opening_element(&mut self, elem: &mut JSXOpeningElement<'a>) {
-        if self.options.is_jsx_self_plugin_enabled() {
+    pub fn transform_jsx_opening_element(
+        &mut self,
+        elem: &mut JSXOpeningElement<'a>,
+        ctx: &TraverseCtx<'a>,
+    ) {
+        let is_constructor = ctx
+            .find_scope(|scope| {
+                if scope.is_empty() || scope.is_arrow() {
+                    return FinderRet::Continue;
+                }
+                FinderRet::Found(scope.is_constructor())
+            })
+            .unwrap_or(false);
+
+        if (!is_constructor
+            || ctx
+                .find_ancestor(|ancestor| match ancestor {
+                    Ancestor::ClassBody(class) => FinderRet::Found(class.super_class()),
+                    _ => FinderRet::Continue,
+                })
+                .is_none())
+            && self.options.is_jsx_self_plugin_enabled()
+        {
```

Comment:
> On the scopes logic itself, it looks _almost_ correct to me. But...
> 
> 1. What about if we're not in a function at all? Then `__self` shouldn't be added. I guess this means checking for `ScopeFlags::Top` and making `is_constructor = true` in that case (and rename it `is_constructor_or_top_level`).

You are right. This is the case because we are currently writing the plugin based on Babel's test cases for the current lack of such tests so I haven't done anything based on that yet

Reply:
Should we add `ScopeFlags::Function` for MethodDefinition? Like what VisitMut did?

Reply:
> Should we add `ScopeFlags::Function` for MethodDefinition? Like what VisitMut did?

Yes I think we should. But from what I can see, `Visit` and `VisitMut` use `ScopeFlags::empty()` for class methods. If we change this in `Traverse`, we should change it in `Visit` and `VisitMut` too, right?

https://github.com/oxc-project/oxc/blob/6fd7a3c42712487430ddfd627f61654a32ab58ce/crates/oxc_ast/src/visit/visit.rs#L1379-L1390

https://github.com/oxc-project/oxc/blob/6fd7a3c42712487430ddfd627f61654a32ab58ce/crates/oxc_ast/src/visit/visit_mut.rs#L1416-L1431

Reply:
Ah no, now I get it! I missed some logic. `Visit::walk_function` sets the `ScopeFlags::Function` bit. This is my mistake - I translated the logic wrongly in Traverse. I'll fix it now.

Reply:
#3277

Reply:
> If you can merge that and then rebase this PR on top, then I'll review again. I do think there's still one problem remaining (but I've been wrong about everything else so far, so maybe not!?)

merged

Reply:
I'm going to resolve this conversation and start again on the up-to-date code.